libevmasm/Assembly.cpp

```diff
@@ -784,6 +785,15 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
 				return _i == AssemblyItem{Instruction::MSIZE} || _i.type() == VerbatimBytecode;
 			});
 
+			auto runGas = [&](AssemblyItems const& items, KnownState _state) {
+				GasMeter gasMeter{std::make_shared<KnownState>(_state), _settings.evmVersion};
```

Comment:
Are you sure you want to make a copy of the whole `KnownState` here? Two copies in fact, because `make_shared()` will allocate another one.

```suggestion
			auto runGas = [&](AssemblyItems const& items, std::shared_ptr<KnownState> _state) {
				GasMeter gasMeter{_state, _settings.evmVersion};
```

Also, `getKnownState()` should probably be returning `shared_ptr` as well.